Refactor Docs for Main and Transactions (#116)
evmos/evmosjs: README.md

# evmosjs

JS and TS libs for Evmos.

## Installation

> evmosjs uses [buf.build](https://buf.build/) to manage Protobuf dependencies. To install evmosjs packages in your project,

Reviewer: ```suggestion
evmosjs uses [buf.build](https://buf.build/) to manage [Evmos Protobuf dependencies](https://buf.build/evmos). To install evmosjs packages in your project,
```

follow the instructions corresponding to your package manager.

### NPM

Add the following line to an `.npmrc` file in your project root:

> ```ini

Reviewer: is `ini` a valid option?

Author: Yes, that's the format of the `.npmrc` file:
https://docs.npmjs.com/cli/v9/configuring-npm/npmrc#global-config-file

GitHub uses Linguist to parse MarkDown:
https://github.com/github/linguist

Which supports Ini:
https://github.com/github/linguist/blob/master/lib/linguist/languages.yml

Most text editors should also support it by way of Highlight.js:
https://github.com/highlightjs/highlight.js/blob/main/SUPPORTED_LANGUAGES.md

@buf:registry=https://buf.build/gen/npm/v1
```

> Then run `npm install evmosjs` (or `npm install @evmosjs/[package]`).

Reviewer: nit, prefer code blocks for installation instructions since they are easier to copy

### Yarn v2.x or v3.x

Add the following to an `.yarnrc.yml` file in your project root:

```yaml
npmScopes:
  buf:
    npmRegistryServer: "https://buf.build/gen/npm/v1"
```

> Then run `yarn add evmosjs` (or `yarn add @evmosjs/[package]`).

Reviewer: ditto

> Note that Yarn v1 is not supported ([why not?](https://docs.buf.build/bsr/remote-packages/npm#other-package-managers)).

Reviewer: ```suggestion
Note that Yarn v1 is not supported ([see explanation](https://docs.buf.build/bsr/remote-packages/npm#other-package-managers)).
```

  Chain,
  Sender,
  Fee,
  TxContext,
  MsgSendParams,
  createTxMsgSend,
  TxPayload,
} from '@evmos/transactions'

const chain: Chain = {
  chainId: 9001,
  cosmosChainId: 'evmos_9001-2',
}

// Populate the transaction sender parameters using the
// query API.
const sender: Sender = {
  accountAddress: [sender_account_address],
  sequence: [sender_sequence],
  accountNumber: [sender_account_number],
  // Use an empty string if the pubkey is unknown.
  pubkey: [sender_pub_key], 
}

const fee: Fee = {
  amount: '3000000000000000',
  denom: 'aevmos',
  gas: '200000',
}

const memo = ''

const context: TxContext = {
  chain,
  sender,
  fee,
  memo,
}

const params: MsgSendParams = {
  destinationAddress: [destination_address],
  amount: [transaction_amount],
  denom: 'aevmos',
}

const tx: TxPayload = createTxMsgSend(context, params)
```

### Sign the Transaction with MetaMask

Evmos supports EIP-712 signatures for Cosmos payloads to be signed using Ethereum wallets such as MetaMask.

```ts
import { createTxRaw } from '@evmos/proto'
import { evmosToEth } from '@evmos/address-converter'

// First, populate a TxContext object and create a signable Tx payload.
const context = ...

> const tx = ...

Reviewer 2: Nit: define these

const { sender } = context

// Initialize MetaMask and sign the EIP-712 payload.
await window.ethereum.enable()

const senderHexAddress = evmosToEth(sender.accountAddress)
const eip712Payload = JSON.stringify(tx.eipToSign)

const signature = await window.ethereum.request({
  method: 'eth_signTypedData_v4',
  params: [senderHexAddress, eip712Payload],
})

// Create a signed Tx payload that can be broadcast to a node.
const signatureBytes = Buffer.from(signature.replace('0x', ''), 'hex')

const { signDirect } = tx
const bodyBytes = signDirect.body.toBinary()
const authInfoBytes = signDirect.authInfo.toBinary()

const signedTx = createTxRaw(
  bodyBytes,
  authInfoBytes,
  [signatureBytes],
)
```

### Sign the Transaction with Keplr (SignDirect)

EvmosJS supports Cosmos SDK `SignDirect` payloads that can be signed using Keplr.

```ts
import { createTxRaw } from '@evmos/proto'

// First, populate a TxContext object and create a signable Tx payload.
const context = ...

> const tx = ...

Reviewer 2: Nit: define these

const { chain, sender } = context
const { signDirect } = tx

const signResponse = await window?.keplr?.signDirect(
  chain.cosmosChainId,
  sender.accountAddress,
  {
    bodyBytes: signDirect.body.toBinary(),
    authInfoBytes: signDirect.authInfo.toBinary(),
    chainId: chain.cosmosChainId,
    accountNumber: new Long(sender.accountNumber),
  },
)

if (!signResponse) {
  // Handle signature failure here.
}

const signatures = [
  new Uint8Array(Buffer.from(signResponse.signature.signature, 'base64')),
]

const { signed } = signResponse

const signedTx = createTxRaw(
  signed.bodyBytes,
  signed.authInfoBytes,
  signatures,
])
```

### Sign the Transaction with Keplr (EIP-712)

> EvmosJS also supports signing EIP-712 payloads using Keplr. This is necessary for Ledger users on Keplr, since the Ledger cannot sign SignDirect payloads.

Reviewer: ```suggestion
EvmosJS also supports signing [EIP-712](https://eips.ethereum.org/EIPS/eip-712) payloads using Keplr. This is necessary for Ledger users on Keplr, since the Ledger device cannot sign `SignDirect` payloads.
```

```ts
import { EthSignType } from '@keplr-wallet/types';
import { createTxRaw } from '@evmos/proto'

// First, populate a TxContext object and create a signable Tx payload.
const context = ...

> const tx = ...

Reviewer 2: Nit: ditto
